NEXT-16508 - Fixed entity exists and not validators work correctly with composite constraints.

AndreasA opened the pull request:
### 1. Why is this change necessary?

To allow entity exists and not exists validators to be used in combination with composite constraints like Symfony's `All` constraint. 

For more information, see: https://issues.shopware.com/issues/NEXT-16508


### 2. What does this change do, exactly?
Fixes https://issues.shopware.com/issues/NEXT-16508

### 3. Describe each step to reproduce the issue or behaviour.
See https://issues.shopware.com/issues/NEXT-16508

### 4. Please link to the relevant issues (if any).
https://issues.shopware.com/issues/NEXT-16508

### 5. Checklist

- [X] I have written tests and verified that they fail without my change
- [X] I have squashed any insignificant commits
- [X] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [X] I have written or adjusted the documentation according to my changes
- [X] This change has comments for package types, values, functions, and non-obvious lines of code
- [X] I have read the contribution requirements and fulfil them.

AndreasA committed:
Fixed NEXT-16508

Dominik28111 commented:
Hi @AndreasA , 

I've just set your PR to “Scheduled”.
This means our initial screening is passed.
Your PR will now be reviewed and tested before being put trough our internal extended testsuite and review process.
We will keep you informed wether something has to be done by you or when the PR is merged : )
If you have any questions feel free to ask.

AndreasA commented:
@Dominik28111 Shall I rebase the latest trunk into it because I think the cypress tests have been fixed there or can I ignored this?

Dominik28111 commented:
@AndreasA This is not what is described in the ticket and what you have fixed or?
Because actually it is described there "Hey if you have several entity exists, then the criteria is cloned again and again and then it does not work"... and you check here only whether entity exists works if the entities are there and once if not.

AndreasA commented:
@Dominik28111

Hi,

no it fixes exactly what I described. Maybe my description was not good enough though. The cloning was missing before, however, it is necessary for the validator to work if used multiple times. It also is necessary that the original criteria is not changed.

I fixed the EntityExistsValidator and EntityNotExistsValidator to be capable to be used in combination with Symfony's composite validators like "All" which this does.

In the EntityExistsValidator test, I check if multiple entity IDs exist by using Symfony's "All" constraint and an array of IDs.

```php
...

  new All(
                [
                    'constraints' => [
                        new EntityExists(
                            ['context' => $context, 'entity' => LocaleDefinition::ENTITY_NAME]
                        ),
                    ],
                ]
            )


...
        $violations = $validator->getViolations(['ids' => [$id1, $id2]], $definition);
```

Without my change this will fail as previously the `criteria` object was not cloned which resulted in a query like this for "id2":
`id = $id1 AND id = $id2` and that always failed and as I use `All` and not AtLeastOne or something like this it produces the wrong result.

if you now check for 
```php
        $violations = $validator->getViolations(['ids' => [$id1, $id2, `Uuid::randomHex(]],` $definition);
```

then it will still fail as the third ID does not exist as it is a new UUID.


Additionally, I fixed that EntityNotExistsValidator did not work at all. It was only true if no entities at all `xisted` for that table.
Furthermore, I added a limit to `=1` so that if one were to use it like primary property `active` and `true`  it will not resolve all entities but just one as that is enough for existence (though it is an unlikely usecase but the change is helpful anyway).

AndreasA commented:
@Dominik28111 

But I can add an additional check like this to verify that scenario works too:

```
$validator->getViolations(['id' => $id1], $definition);
$validator->getViolations(['id' => $id2], $definition);
```

AndreasA commented:
@Dominik28111 I have now added two more tests that also verify using the same constraint multiple times.

AndreasA commented:
@Dominik28111 Let me know if this is fine now or if you need anything else ?

AndreasA commented:
@Dominik28111 Sorry for one more comment.

I now added various additional tests that show the issues from before and the fix.
One that checks that a provided criteria object is not modified and one that checks that if the same validation definition is used it will work as expected now.

If you try my tests without the validator changes they will all fail but of course they should not fail as e.g. one shows that entit exists will always fail currently if used more than once.

Let me know if everything is fine now

AndreasA commented:
@Dominik28111 I have improved the tests now to not require the DataValidator but use Symfony's validator directly that way an unnecessary layer for these tests is removed. Otherwise nothing has chnaged to the comment just before this one.

Dominik28111 commented:
Thanks for your contribution. Your PR will be merged when the pipeline is green.

AndreasA commented:
@Dominik28111 The pipeline is not green because the trunk pipeline isn't either: https://github.com/shopware/platform/runs/3281261644

The PHPUnit tests etc. are working. the cypress admin tests are failing but as mentioned those fail for the trunk too, when will those be fixed?

mitelg commented:
I'm sure @Dominik28111 is talking about our internal pipeline 😉

AndreasA commented:
@mitelg @Dominik28111 Ah that makes more sense 😄 and I see it is already merged.
